From 3d60280c1c16265340ae4b7a0b90aa04f32ce5d2 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 15 Apr 2024 10:47:39 +0200
Subject: [PATCH 0454/1695] scpfw: module/optee/voltd_regulator: Upgrade
 voltage list API

Upgrade OPTEE_VOLTD_REGULATOR module to OP-TEE latest API functions
to retrieve the list of the supported voltages. OP-TEE regulator
framework API changed (see Link tag) to fix an implementation issue
reported by Clang.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 61b74eeddac424063e7a8b795cee339fd53be888 scpfw: module/optee/voltd_regulator: Upgrade voltage list API]
Change-Id: Ia24ad713be6aeb89637409f76137602bd788d0ca
Link: https://github.com/OP-TEE/optee_os/pull/6487
Fixes: bf5b4150ff78 ("module/optee: Add voltage domain regulators")
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../src/mod_optee_voltd_regulator.c           | 26 ++++++++++++-------
 1 file changed, 16 insertions(+), 10 deletions(-)

diff --git a/core/lib/scmi-server/SCP-firmware/module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c b/core/lib/scmi-server/SCP-firmware/module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c
index b1b970f3b..a329c82e2 100644
--- a/core/lib/scmi-server/SCP-firmware/module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c
+++ b/core/lib/scmi-server/SCP-firmware/module/optee/voltd_regulator/src/mod_optee_voltd_regulator.c
@@ -212,9 +212,12 @@ static int optee_voltd_regulator_get_info(fwk_id_t dev_id,
         info->level_range.min_uv = 0;
         info->level_range.max_uv = 0;
     } else {
-        struct regulator_voltages *desc;
+        struct regulator_voltages_desc *desc;
+        const int *levels;
+        TEE_Result res;
 
-        if (regulator_supported_voltages(ctx->regulator, &desc) != TEE_SUCCESS) {
+        res = regulator_supported_voltages(ctx->regulator, &desc, &levels);
+        if (res != TEE_SUCCESS) {
             return FWK_E_SUPPORT;
         }
 
@@ -222,14 +225,14 @@ static int optee_voltd_regulator_get_info(fwk_id_t dev_id,
             info->name = regulator_name(ctx->regulator);
             info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
             info->level_range.level_count = desc->num_levels;
-            info->level_range.min_uv = desc->entries[0];
-            info->level_range.max_uv = desc->entries[desc->num_levels - 1];
+            info->level_range.min_uv = levels[0];
+            info->level_range.max_uv = levels[desc->num_levels - 1];
         } else if (desc->type == VOLTAGE_TYPE_INCREMENT) {
             info->name = regulator_name(ctx->regulator);
             info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
-            info->level_range.min_uv = desc->entries[0];
-            info->level_range.max_uv = desc->entries[1];
-            info->level_range.step_uv = desc->entries[2];
+            info->level_range.min_uv = levels[0];
+            info->level_range.max_uv = levels[1];
+            info->level_range.step_uv = levels[2];
         } else {
             fwk_unreachable();
             return FWK_E_PANIC;
@@ -250,7 +253,9 @@ static int optee_voltd_regulator_level_from_index(fwk_id_t dev_id,
                                                   int32_t *level_uv)
 {
     struct optee_voltd_regulator_dev_ctx *ctx;
-    struct regulator_voltages *desc;
+    struct regulator_voltages_desc *desc;
+    const int *levels;
+    TEE_Result res;
 
     ctx = get_ctx(dev_id);
     if (ctx == NULL) {
@@ -268,7 +273,8 @@ static int optee_voltd_regulator_level_from_index(fwk_id_t dev_id,
         return FWK_SUCCESS;
     }
 
-    if (regulator_supported_voltages(ctx->regulator, &desc) != TEE_SUCCESS) {
+    res = regulator_supported_voltages(ctx->regulator, &desc, &levels);
+    if (res != TEE_SUCCESS) {
         return FWK_E_SUPPORT;
     }
 
@@ -279,7 +285,7 @@ static int optee_voltd_regulator_level_from_index(fwk_id_t dev_id,
         return FWK_E_RANGE;
     }
 
-    *level_uv = (int32_t)desc->entries[index];
+    *level_uv = levels[index];
 
     FWK_LOG_DEBUG(
         MOD_PREFIX "Get level from index for %u/%s: index %u, level %"PRId32"uV",
-- 
2.47.2

